### PR TITLE
[fix](merge-cloud) Fix drop rollup/mv leak meta kvs and data objects resource

### PR DESCRIPTION
**Pull request by swjtu-zhanglei**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by SWJTU-ZhangLei** [comment]
run buildall

**Comment by SWJTU-ZhangLei** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 1]
```diff
@@ -701,6 +701,32 @@ public void dropMaterializedIndex(long tableId, List<Long> indexIds, boolean dro
         }
     }
 
+    /**
+     * for cloud mode, drop rollupIndex/materializedIndexId in kv meta store
+     * @param tableId
+     * @param materializedIndexId
+     */
+    public void eraseIndexDropBackendReplicas(long tableId, List<Long> indexIdList) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:709]: This is a cofusing name, consider `eraseDroppedIndex`

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 2]
```diff
@@ -701,6 +701,32 @@ public void dropMaterializedIndex(long tableId, List<Long> indexIds, boolean dro
...
+        if (indexIdList == null || indexIdList.size() == 0) {
+            LOG.warn("indexIdList is empty");
+            return;
+        }
+        long tryTimes = 1;
+        while (true) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:715]: avoid using an infinite loop.
try hundreds of times and return with an error log.

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java (hunk 2 above)
Comment (SWJTU-ZhangLei) [reply fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:715]: > avoid using an infinite loop. try hundreds of times and return with an error log.

may cause resource leak, are you sure?

Comment (SWJTU-ZhangLei) [reply fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:715]: > avoid using an infinite loop. try hundreds of times and return with an error log.

resolved

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java (hunk 1 above)
Comment (SWJTU-ZhangLei) [reply fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java]: > This is a cofusing name, consider `eraseDroppedIndex`

resolved

**Comment by SWJTU-ZhangLei** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 3]
```diff
@@ -701,6 +701,37 @@ public void dropMaterializedIndex(long tableId, List<Long> indexIds, boolean dro
         }
     }
 
+    /**
+     * for cloud mode, drop rollup/materializedIndex in kv meta store
+     * @param tableId
+     * @param indexIdList
+     */
+    public void eraseDroppedIndexBackendReplicas(long tableId, List<Long> indexIdList) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:709]:
```suggestion
    public void eraseDroppedIndex(long tableId, List<Long> indexIdList) {
```

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 4]
```diff
@@ -701,6 +701,37 @@ public void dropMaterializedIndex(long tableId, List<Long> indexIds, boolean dro
...
+                LOG.warn("failed to drop index {} of table {}, try cnt {} reaches maximum retry count",
+                            indexIdList, tableId, tryCnt);
+                break;
+            }
+
+            try {
+                dropMaterializedIndex(tableId, indexIdList, false);
+                break;
+            } catch (Exception e) {
+                LOG.warn("tryCnt:{}, eraseDroppedIndexBackendReplicas exception:", tryCnt, e);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:726]:
```suggestion
                LOG.warn("tryCnt:{}, eraseDroppedIndex exception:", tryCnt, e);
```

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 5]
```diff
@@ -701,6 +701,37 @@ public void dropMaterializedIndex(long tableId, List<Long> indexIds, boolean dro
...
+            }
+            sleepSeveralMs();
+        }
+
+        LOG.info("eraseDroppedIndexBackendReplicas finished, tableId:{}, indexIdList:{}",
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:731]:
```suggestion
        LOG.info("eraseDroppedIndex finished, tableId:{}, indexIdList:{}",
```

File: fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java [hunk 6]
```diff
@@ -1023,6 +1023,10 @@ public void erasePartitionDropBackendReplicas(List<Partition> partitions) {
         // no need send be delete task, when be report its tablets, fe will send delete task then.
     }
 
+    public void eraseDroppedIndexBackendReplicas(long tableId, List<Long> indexIdList) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java:1026]:
```suggestion
    public void eraseDroppedIndex(long tableId, List<Long> indexIdList) {
```

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 7]
```diff
@@ -942,15 +944,18 @@ public void processBatchDropRollup(List<AlterClause> dropRollupClauses, Database
             long tableId = olapTable.getId();
             String tableName = olapTable.getName();
             editLog.logBatchDropRollup(new BatchDropInfo(dbId, tableId, tableName, indexIdSet));
+            deleteIndexList = indexIdSet.stream().collect(Collectors.toList());
             LOG.info("finished drop rollup index[{}] in table[{}]",
                     String.join("", rollupNameSet), olapTable.getName());
         } finally {
             olapTable.writeUnlock();
         }
+        Env.getCurrentInternalCatalog().eraseDroppedIndexBackendReplicas(olapTable.getId(), deleteIndexList);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:953]:
```suggestion
        Env.getCurrentInternalCatalog().eraseDroppedIndex(olapTable.getId(), deleteIndexList);
```

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 8]
```diff
@@ -973,6 +979,7 @@ public void processDropMaterializedView(DropMaterializedViewStmt dropMaterialize
...
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:982]:
```suggestion
        Env.getCurrentInternalCatalog().eraseDroppedIndex(olapTable.getId(), deleteIndexList);
```

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 9]
```diff
@@ -1060,6 +1067,10 @@ public void replayDropRollup(DropInfo dropInfo, Env env) throws MetaNotFoundExce
         } finally {
             olapTable.writeUnlock();
         }
+
+        List<Long> deleteIndexList = new ArrayList<Long>();
+        deleteIndexList.add(rollupIndexId);
+        Env.getCurrentInternalCatalog().eraseDroppedIndexBackendReplicas(olapTable.getId(), deleteIndexList);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:1073]:
```suggestion
        Env.getCurrentInternalCatalog().eraseDroppedIndex(olapTable.getId(), deleteIndexList);
```

**Comment by dataroaring** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**